_posts/2020-11-02-dynamically-changeable-logging.md

```diff
@@ -0,0 +1,163 @@
+
+`Log4j2` implementation supports dynamic reloading of logging configuration from properties file.
+`Log4j` doesn't support reloading the whole log configuration from properties file, but it can be changed programmatically using the Log4j APIs.
+And that is why each implementation requires a different approach, which we will look at now.
+
+### Strimzi Cluster Operator
+
+In order to change the Cluster Operator logging config, a simple change is required.
+You have to configure the logging settings in the `ConfigMap` used for Cluster Operator configuration.
+This `ConfigMap` is by default called `strimzi-cluster-operator`,
+To edit the logging configuration use
+```
+$ kubectl edit cm strimzi-cluster-operator
+```
+
+Under the field `log4j2.properties` you should see current logging configuration.
+Now you can edit this configuration to whatever you want with one exception.
+
+The entry `monitorInterval` must stay in place.
+When it is removed, the configuration will not be reloaded dynamically anymore.
+To apply changes in such a case, the pod needs to be rolled. 
+
+The second way to change `ConfigMap` is to modify the file `install/cluster-operator/050-ConfigMap-strimzi-cluster-operator.yaml`, and apply the changes by this command:
+
+```
+$ kubectl apply -f install/cluster-operator/050-ConfigMap-strimzi-cluster-operator.yaml
+```
+
+The same restrictions apply.
+The entry `monitorInterval` must stay in place.
+
+The change in logging takes some time to be applied. 
+You can find out more in the implementation details section of this post.
+
+### Topic Operator, User Operator and Kafka resources
+
+The approach to changing the logging configuration of User Operator, Topic Operator, Kafka Bridge, Kafka brokers, Kafka Connect and Kafka MirrorMaker2 did not change.
+The logging is still configurable via `inline` or `external` logging section of custom resource specification.
+
+Example of `inline` logging.
+```yaml
+...
+spec:
+  logging:
+    type: inline
+      loggers:
+        kafka.root.logger.level = INFO
+...
+```
+
+Example of external logging.
+```yaml
+...
+spec:
+  logging:
+    type: external
+      name: my-external-config-map
+...
+```
+
+Only loggers can be changed dynamically.
+Logging appenders are changed using an external `ConfigMap` and the rolling update is triggered.
+
+This does not apply for Kafka Bridge, User operator, Topic operator and Cluster operator.
+Because these use `log4j2` framework, even logging appenders are changeable dynmaicaly.
+
+
+## Verifying logging levels are set correctly
+
+In most cases, it takes a few seconds to change the logging level, and then you can observe the change in the logs.
+If you want to check the loggers and logging levels are changed, you can use these commands.
+
+For Kafka brokers:
+
+```
+$ kubectl  exec <kafka-cluster-name>-kafka-0 -- bin/kafka-configs.sh --bootstrap-server <kafka-cluster-name>-kafka-bootstrap:9092 --entity-type broker-loggers --entity-name 0 --describe
```

Comment:
what's this `--entity-name 0` ?

Reply:
It is the name of the broker. We can say it is `broker.id` from the server properties.